constructors.md

```diff
@@ -1 +1,112 @@
 # Constructors
+
+Assume that `DType` is the same as `gdf_dtype`, and that `device_buffer` is effectively a `rmm::device_vector` and `host_buffer` is effectively a `std::vector`. 
+
+Note: This design document purposefully ignores the issue of `allocators` and controlling how the device memory for the column is allocated. This will be added as the allocator design in fleshed out. For the timebeing, assume that an allocator handle can be passed into each of the constructors that controls how the device memory for the data and bitmask is allocated.
+
+```
+/**---------------------------------------------------------------------------*
+ * @brief Construct a new column from a size, type, and option to
+ * allocate bitmask.
+ * 
+ * Both the data and bitmask are unintialized.
+ *
+ * @param[in] size The number of elements in the column
+ * @param[in] type The element type
+ * @param[in] allocate_bitmask Optionally allocate an appropriate sized
+ * bitmask
+ *---------------------------------------------------------------------------**/
+column(int size, DType type, bool allocate_bitmask = false);
+
+/**---------------------------------------------------------------------------*
+ * @brief Construct a new column from a type, and device_buffers for
+ * data and bitmask that will be *deep* copied.
+ *
+ * @param[in] dtype The element type
+ * @param[in] data_buffer device_buffer whose data will be *deep* copied
+ * @param[in] mask_buffer Optional device_buffer whose data will be *deep*
+ *copied
+ *---------------------------------------------------------------------------**/
+column(DType dtype, device_buffer data,
```

Comment:
The copy-constructor for `device_buffer` will deep-copy the associated device memory. So this constructor passed `device_buffer` by copy if a deep-copy is desired.

If a shallow-copy is desired, this constructor may be used:

```
column(DType dtype, device_buffer&& data, device_buffer&& mask_buffer = device_buffer{});
```

The `&&` is an "r-value reference" which means it only accepts r-values.

So you could use it like this:

```
device_buffer A;
device_buffer B:

column( GDF_INT32, A); // deep copies A's data
column( GDF_INT32, std::move(B)); // B's data is _moved_ from B and B is no longer valid to use
column(GDF_INT32, device_buffer{...}); // You create a temporary and pass that in, and it will shallow copy the temporaries data
```

Reply:
Ah, got it.  Silly me, I thought the column constructor's body would be performing the deep copy explicitly, but yeah that makes sense.